### PR TITLE
Test newest jsonattrs lib

### LINKED ISSUE #1656: API 500 Error - "is not JSON serializable"
### Steps to reproduce the error

The below error was caused by the QGIS plugin when it uploaded joined tables. I've logged a bug for it in the QGIS repo, but we still shouldn't give a 500 error. Is there a way to make this serializable instead?

TypeError at /api/v1/organizations/chickenville/projects/kate-magical-test-file-2/spatial/
{'have_int': '4', 'name': 'blah blah blah', 'have_dec': '2.100000', 'fake-date': '2017-01-01', 'chickenville/kate-magical-test-file-2/relationships_rel_id': 'x4qc56zp8ybzj5gkxuejinku', 'chickenville/kate-magical-test-file-2/parties_name': 'two', 'chickenville/kate-magical-test-file-2/relationships_rel_name': 'CU', 'chickenville/kate-magical-test-file-2/relationships_party_id': 'awajdc97ktpve6gj6bqrjvwe', 'chickenville/kate-magical-test-file-2/parties_type': 'IN'} is not JSON serializable

### PR DESCRIPTION
**Pull request by alukach**
#### Why I made this change

Fixes #1656 

#### Description of the change

* Update `django-jsonattrs` to not attempt to validate the contents of a `jsonattrs` against that field's schema when reading from the field.  Instead, we will only scrutinize the data when insterting it into the field.
* Has the bonus effect of improving response times for API endpoints that include `jsonattrs` fields.

##### Failure without fix

On `master` branch, in shell:

```
$ DJANGO_SETTINGS_MODULE=config.settings.dev_debug ./manage.py shell_plus`)
```

Use the `cpaste` command and paste:

```py
import json

from django.db import connection
from rest_framework.renderers import JSONRenderer

from organization.models import Project
from spatial.serializers import SpatialUnitSerializer

# Let's force insert some bad data into our DB
_id = "asd"
with connection.cursor() as c:
    sql = "INSERT INTO {table} ({columns}) VALUES ({values})"
    data = [
        ('id', repr("asdf")),
        ('project_id', repr(Project.objects.first().id)),
        ('type', "'PA'"),
        ('geometry', "null"),
        ('area', '0'), 
        ('attributes', repr(json.dumps({"foo": 'bar'}))), 
        ('created_date', "'2017-01-01T01:01:59'"),
        ('last_updated', "'2017-01-01T01:01:59'"),
    ]
    sql = sql.format(
        table=SpatialUnit._meta.db_table,
        columns=', '.join([d for d, _ in data]),
        values=', '.join([d for _, d in data]),
    )
    c.execute(sql)

# This is going to fail
l = SpatialUnit.objects.get(id="asdf")
data = SpatialUnitSerializer(l).data
JSONRenderer().render(data)
```

##### With fix

1. Checkout `update-django-jsonattrs` branch
1. `pip install -r requirements.txt`
1. In `shell_plus`:

```
# This won't fail
l = SpatialUnit.objects.get(id="asdf")
data = SpatialUnitSerializer(l).data
JSONRenderer().render(data)
```

### When should this PR be merged

Cadasta/django-jsonattrs#39 and Cadasta/django-jsonattrs#30 should both be merged and a new version released for `django-jsonattrs`. This PR should be closed and #1987 should be updated to use the new version of `django-jsonattrs`.

### Risks

~~I have faith in our unit and functional tests, so I feel that risks are low.~~

**Please review [this comment](https://github.com/Cadasta/cadasta-platform/pull/1984#issuecomment-366141625) and let me know if it seems sensible**

#### General

**Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 

- [ ] Review 1
- [ ] Review 2

**Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 

- [ ] Review 1
- [ ] Review 2

**Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

- [ ] Review 1
- [ ] Review 2

#### Functionality

**Are all requirements met?** Compare implemented functionality with the requirements specification.

- [ ] Review 1
- [ ] Review 2

**Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

- [ ] Review 1
- [ ] Review 2

#### Code

**Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.

- [ ] Review 1
- [ ] Review 2

**Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 

- [ ] Review 1
- [ ] Review 2

**Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

- [ ] Review 1
- [ ] Review 2

**Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).

- [ ] Review 1
- [ ] Review 2

**Has the scalability of this change been evaluated?**

- [ ] Review 1
- [ ] Review 2

**Is there a maintenance plan in place?**

- [ ] Review 1
- [ ] Review 2

#### Tests

**Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 

- [ ] Review 1
- [ ] Review 2

**If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.

- [ ] Review 1
- [ ] Review 2

**If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

- [ ] Review 1
- [ ] Review 2

#### Security

**Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**

- [ ] Review 1
- [ ] Review 2

**Are all UI and API inputs run through forms or serializers?** 

- [ ] Review 1
- [ ] Review 2

**Are all external inputs validated and sanitized appropriately?**

- [ ] Review 1
- [ ] Review 2

**Does all branching logic have a default case?**

- [ ] Review 1
- [ ] Review 2

**Does this solution handle outliers and edge cases gracefully?**

- [ ] Review 1
- [ ] Review 2

**Are all external communications secured and restricted to SSL?**

- [ ] Review 1
- [ ] Review 2

#### Documentation

**Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).

- [ ] Review 1
- [ ] Review 2

**Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).

- [ ] Review 1
- [ ] Review 2

**Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 

- [ ] Review 1
- [ ] Review 2

### CODE REVIEW

File: cadasta/party/tests/test_models.py [hunk 1]
```diff
@@ -231,6 +231,15 @@ def test_project_reverse_tenure_relationships(self):
...
     def test_set_attributes(self):
+        # add attribute schema
+        content_type = ContentType.objects.get(
+            app_label='party', model='tenurerelationship')
+        sch = Schema.objects.create(content_type=content_type, selectors=())
+        attr_type = AttributeType.objects.get(name="text")
+        Attribute.objects.create(
+            schema=sch, name='description', long_name='Description',
+            required=False, index=1, attr_type=attr_type
+        )
```
Comment (alukach) [inline cadasta/party/tests/test_models.py]: This allows the test to pass.  I have no idea how the test ever passed before, I feel like [the `KeyError` that occurs without this change](https://travis-ci.org/Cadasta/cadasta-platform/jobs/342080311#L1013-L1059) should have alway occurred as there's no related `Attribute` to allow that attribute to be set on the `TenureRelationship` object.

I copied this code from an earlier test: https://github.com/Cadasta/cadasta-platform/pull/1984/commits/d0f68bbab7926a3bb88e26c41c662b458d21e72c#diff-651e8c2403a2188377ef4dddb4a740cbR162

**Comment by alukach** [comment]
As mentioned [here](https://github.com/Cadasta/cadasta-platform/pull/1984/files#r168672071), I don't think the tests should have ever passed.  Basically, what was happening was that when `tenurerelationship.save()` was called:
https://github.com/Cadasta/cadasta-platform/blob/d16b4d2f8a14fb4b77fd577026d85d8443e5c1a7/cadasta/party/tests/test_models.py#L233-L242

[`self.setup_from_dict()` is called with `self.db_val`](https://github.com/Cadasta/django-jsonattrs/blob/9e7ee0a8e6937d95af2675365237f9c59a80a765/jsonattrs/fields.py#L62-L64):
```py
    def _pre_save_selector_check(self, strict=False):
        if not self._setup:
            self.setup_from_dict(self._db_val)
```

`self._db_val == {}`

During setup, no related schemas are found for the `tenurerelationship` instance and so we [skip the check on the data being inputted](https://github.com/Cadasta/django-jsonattrs/blob/9e7ee0a8e6937d95af2675365237f9c59a80a765/jsonattrs/fields.py#L27-L30):

```py
    def setup_from_dict(self, dict):
        self.setup_schema()
        if dict is None or len(dict) == 0:
            self._setup = False
        # ...
```

I think this was a mistake and seems like a probable way that much of the bad data makes its way into our `jsonattrs` columns of our database.

**Comment by seav** [comment]
> I think this was a mistake and seems like a probable way that much of the bad data makes its way into our `jsonattrs` columns of our database.

So there is still a possible bug in the django-jsonattrs library? Maybe this `dict is None or len(dict) == 0` check should only be done if we are creating the `JSONAttributes` object when reading the model instance from the database and should **not** be checked when trying to create a model instance from user input?

**Comment by oliverroick** [comment]
> Maybe this dict is None or len(dict) == 0 check should only be done if we are creating the JSONAttributes object when reading the model instance from the database and should not be checked when trying to create a model instance from user input?

Sounds like something we could do but I'm not fussed. Leaving it to @alukach to decide.

**Comment by oliverroick** [comment]
I have already merge #1987 with dev-version of `jsonattrs` (0.1.24.dev1). Before we merge this PR we should merge Cadasta/django-jsonattrs#30, release `jsonattrs` 0.1.24 and update the requirements for this branch.

**Comment by alukach** [comment]
@seav 
@oliverroick 

> So there is still a possible bug in the django-jsonattrs library? Maybe this `dict is None or len(dict) == 0` check should only be done if we are creating the `JSONAttributes` object when reading the model instance from the database and should **not** be checked when trying to create a model instance from user input?

Sorry, it was late and I wasn't explaining myself well. I believe this bug is effectively fixed by these lines of Cadasta/django-jsonattrs#30:

```py
    def _pre_save_selector_check(self, strict=False):
        if not self._setup:
            self.setup_from_dict(self._db_val if self._instance._state.adding
                                 else self._get_from_instance())
        # ...
```
https://github.com/Cadasta/django-jsonattrs/blob/dc5ac19f4b993b200af84abe93a44bbacaaf0a6e/jsonattrs/fields.py#L68-L71

With these changes, `setup_from_dict` does a pre-save validation on the `_db_val` if the object is being created for the first time (I believe this is because the `JSONAttributeField` value isn't populated at this time) or validated against the `JSONAttributeField` value if the instance is not new.  I admit that I'm still a bit unsure about exactly how everything is working but my confidence is emboldened by the two codebase' unit tests and our functional tests.

I'll see if I can generate a conclusive unit test sometime before end of sprint.

**Pull request merged**